File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:

    void solve(vector<int>& nums, int index,  set<vector<int>>& res, vector<int>& subset) {
        if(index == nums.size()) {
            res.insert(subset);
            return;
        }

        subset.push_back(nums[index]);
        solve(nums, index + 1, res, subset);
        subset.pop_back();
        solve(nums, index + 1, res, subset);

    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        set<vector<int>> res;
        vector<int> subset;
        sort(nums.begin(), nums.end());
        solve(nums, 0, res, subset);
        vector<vector<int>> result;
        for(auto i: res) {
            result.push_back(i);
        }
        return result;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {

    private void solve(int[] nums, int i, List<List<Integer>> res, List<Integer> subset) {
        if(i == nums.length) {
            res.add(new ArrayList<>(subset));
            return;
        }

        subset.add(nums[i]);
        solve(nums, i+1, res, subset);
        subset.remove(subset.size() - 1);

        int index = i+1;

        while(index < nums.length && nums[index] == nums[index-1]) 
            index++;

        solve(nums, index, res, subset);

    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        Arrays.sort(nums);
        solve(nums, 0, res, subset);
        return res;
    }
}